File: Final_Project/Bullet.cpp
#include "Bullet.h"

const int WIN_X = 1600;
const int WIN_Y = 900;

const double PI = 3.1415926535897932384626433832795028841971;

Bullet::Bullet(){
    _x = 0;
    _y = 0;
    _width = 0;
    _height = 0;
    _color = sf::Color::White;
    _angle = 0;
    _speed = 0;
}

Bullet::Bullet(const double xPos, const double yPos, const int width, const int height, const sf::Color color, const double angle, const double speed){
    _x = xPos;
    _y = yPos;
    _width = width;
    _height = height;
    _color = color;
    _angle = angle;
    _speed = speed;
}

void Bullet::setAngle(const double angle){
    _angle = angle;
}

bool Bullet::collide(MovingObject* object){

    if(_x+_width > object->getX() && _x < object->getX() + object->getWidth()){
        if(_y+_height > object->getY() && _y < object->getY() + object->getHeight()){
            return true;
        }
    }
    return false;
}

void Bullet::draw(sf::RenderWindow& window){

    //std::cout << _angle << std::endl;

    sf::RectangleShape bullet;
    bullet.setSize( sf::Vector2f(_width, _height) );
    bullet.setFillColor(_color);
    bullet.setOrigin(_width/2, _height);
    bullet.setPosition(_x, _y);
    bullet.setRotation(_angle);
    window.draw(bullet);
}

void Bullet::updatePosition(){
    _x = _x + cos(_angle * (PI/180) - PI/2) * _speed;
    _y = _y + sin(_angle * (PI/180) - PI/2) * _speed;
}

File: Final Project/src/pkgfinal/project/Bullet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package pkgfinal.project;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import javafx.scene.SnapshotParameters;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.paint.Color;

/**
 *
 * @author brinkman.nathaniel23
 */
public class Bullet extends GameElement implements Renderable, Updatable{
    
    private Image bulletImage;
    private ImageView bulletImageView;
    
    private double theta;
    
    
    public Bullet(double xPos, double yPos, Color c, double angle) throws FileNotFoundException{
        super(xPos, yPos, 15, 8, c);
        
        theta = angle;
        
        FileInputStream input = new FileInputStream("D:\\NetBeansProjects\\Final Project\\images\\bullet.png");
        bulletImage = new Image(input);
        bulletImageView = new ImageView();
        bulletImageView.setImage(bulletImage);
    
    }
    

    
    public void onCollision(GameElement element){
        
        //System.out.println("bullet collide");
        
        if(getX()<=element.getX()+element.getWidth() && getX()+getWidth()>=element.getX() && getY()<=element.getY()+element.getHeight() && getY()+getHeight()>=element.getY() && !(element instanceof Bullet) && !(element instanceof Gun)){
            GameState.removeMovable(this);
        }
        
    }
    
    @Override
    public void update() {
        
        super.setX(getX() + 20*Math.cos(Math.toRadians(theta)));
        super.setY(getY() + 20*Math.sin(Math.toRadians(theta)));
        
    }

    @Override
    public void draw(Canvas canvas, Color col) {
        
        bulletImageView.setRotate(theta);
        bulletImageView.setX(getX());
        bulletImageView.setY(getY());
        
        SnapshotParameters params = new SnapshotParameters();
        params.setFill(Color.TRANSPARENT);
        
        bulletImage = bulletImageView.snapshot(params, null);
        
        canvas.getGraphicsContext2D().drawImage(bulletImage, getX(), getY());
        
        
    }
}
